### PR TITLE
Replace TTIR string parsing with structured MLIR walk in Triton kernel  mutation analysis

### PR DESCRIPTION
**Pull request by aakhundov**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #120476

Summary: Previously, we relied on the `lark`-based parsing of the string TTIR representation dumped by the Triton compiler. However, this has proven to be brittle in the face of changes both in the user-written Triton kernel code and in the Triton compiler code.

In this PR, we add an alternative way of mining the function information from the TTIR based on walking the tree of structured MLIR entities. To this end, we rely on the MLIR bindings exposed by `libtriton` (related PR in Triton: https://github.com/openai/triton/pull/3191).

For now, we introduce gating based on whether `ttir_module.hasattr("walk")`. This will allow switching to the newly introduced TTIR analysis approach only when the new MLIR bindings (including that of `ModuleOp::walk`) become available in the Triton pin. Before then, we'll keep using the old string TTIR parsing-based approach.

Test Plan: The new functionality was tested locally with the latest Triton version compiled with the added new MLIR bindings: all Triton kernel mutation tests in `test_triton_kernels.py` are passing. Here we rely on the CI for regression testing, but it won't cover the new functionality due to gating.

Reviewers:

Subscribers:

Tasks:

Tags:

cc @zou3519 @ydwu4

### CODE REVIEW

File: torch/_higher_order_ops/triton_kernel_wrap.py
Comment (oulgen) [inline torch/_higher_order_ops/triton_kernel_wrap.py:173]: you shouldnt need the lambda?
```
defaultdict(defaultdict(list))
```
didn't work?

File: torch/_higher_order_ops/triton_kernel_wrap.py
Comment (oulgen) [inline torch/_higher_order_ops/triton_kernel_wrap.py:179]: Lets add a type to this op to make it clear

File: torch/_higher_order_ops/triton_kernel_wrap.py
Comment (oulgen) [inline torch/_higher_order_ops/triton_kernel_wrap.py:255]: is this how this code gates against tt.reduce? Maybe lets make that explicit, and leave a TODO(oulgen): add support for tt.reduce

**Review by oulgen — COMMENTED** [review]
Nice going!

I think the code looks good but we should not land this until the proposed API lands to triton code base. Lets first make sure Phil&co accept the triton PR.

Also with your custom triton version, please run all of
```
pytest python test/dynamo/test_triton_kernels.py
python test/inductor/test_aot_inductor.py -k test_triton_kernel
```
to make sure it all works

File: torch/_higher_order_ops/triton_kernel_wrap.py
Comment (aakhundov) [reply torch/_higher_order_ops/triton_kernel_wrap.py:173]: I think, the inner `defaultdict(list)` would be an object, whereas the `default_factory` is expected to be a callable.

**Comment by oulgen** [comment]
Should we compare the `functions` from Lark version and this new version to make sure they match? So that we dont accidentally change structures? Not just end to end but compare the dicts. Due to how you handled $2:2 and $2#2 it wont be identical but modulo these it should all be the same.

File: torch/_higher_order_ops/triton_kernel_wrap.py
Comment (aakhundov) [reply torch/_higher_order_ops/triton_kernel_wrap.py:179]: I can add some more type hints, but for exposed bindings this is not trivial (as their types are not visible to mypy), so I'd prefer skipping type hints for `op` and `ttir_module`.

Comment (oulgen) [reply torch/_higher_order_ops/triton_kernel_wrap.py:179]: Would it make sense to add a typehint file for mypy? Similar to our own C++ bindings. Just a suggestion, not a blocker

**Comment by aakhundov** [comment]
> Should we compare the functions from Lark version and this new version to make sure they match? So that we dont accidentally change structures? Not just end to end but compare the dicts. Due to how you handled $2:2 and $2#2 it wont be identical but modulo these it should all be the same.

@oulgen I think, there are more differences than just `tt.reduce` handling: e.g., the Intermediate idx are different between the two representations, I don't create fake intermediates for constant functions, the order of the items in the nested dicts can be different. Plus there is no clear one-to-one  correspondence between the `Op` nodes representing the same TTIR operation in the old and new function dicts: e.g., if there are multiple `tl.load`s, it's not straightforward to relate them in the two representations. I feel, this boils down to verifying isomorphism between the two functions structures which can be tricky.

So, if you feel this is necessary, I can give it a try. But otherwise, perhaps relying on the end-to-end tests using the new functions generation under the hood is sufficient?

**Comment by aakhundov** [comment]
> Also with your custom triton version, please run all of
> pytest python test/dynamo/test_triton_kernels.py
> python test/inductor/test_aot_inductor.py -k test_triton_kernel
> to make sure it all works

@oulgen with the latest version of Triton ([`c681b53`](https://github.com/openai/triton/commit/c681b5390c3a9361480ed9ef4dcc20950dabcbf3)), there are other failures which are unrelated to this PR: e.g., `divisible_by_8` seems to be gone. But I've tested on the current Triton pin + the changes in the MLIR bindings and those tests pass.

**Comment by aakhundov** [comment]
The PR extending MLIR bindings in Triton has been submitted: https://github.com/openai/triton/pull/3191.

File: torch/_higher_order_ops/triton_kernel_wrap.py
Comment (oulgen) [inline torch/_higher_order_ops/triton_kernel_wrap.py:626]:
```suggestion
            log.debug(f"TTIR: \n%s", str(ttir_module))
```

File: torch/_higher_order_ops/triton_kernel_wrap.py
Comment (oulgen) [inline torch/_higher_order_ops/triton_kernel_wrap.py:628]: str of functions would be hard to read, so i'd probably do something like
```suggestion
            for name, function in functions.items():
                log.debug(f"===\t{name}\t===")
                for ret, ops in function.items():
                    log.debug(f"{ret}\t=>\t{ops}")
```

File: torch/_higher_order_ops/triton_kernel_wrap.py
Comment (aakhundov) [reply torch/_higher_order_ops/triton_kernel_wrap.py:626]: Linter complains when trying to use f-strings in log messages.

File: torch/_higher_order_ops/triton_kernel_wrap.py
Comment (aakhundov) [reply torch/_higher_order_ops/triton_kernel_wrap.py:628]: Sure.

**Review by oulgen — APPROVED** [review]

**Pull request closed without merging**
